# Analyze Product Sentiment

In [2]:
import turicreate

# Read product review data

In [38]:
products = turicreate.SFrame('./amazon_baby.sframe')

# Building a sentiment classifier

## Build word count vectors

In [80]:
products['word_count'] = turicreate.text_analytics.count_words(products['review'])

In [41]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'handles': 1.0, 'strippin... |
| {'recommend': 1.0, 'disapp... |
| {'quilt': 1.0, 'the': 1.0,... |
| {'tool': 1.0, 'clever': 1.... |
| {'rock': 1.0, 'many': 1.0,... |
| {'thumb': 1.0, 'or': 1.0, ... |
| {'for': 1.0, 'barnes': 1.0... |
| {'right': 1.0, 'because': ... |
| {'like': 1.0, 'and': 1.0, ... |
| {'in': 1.0, 'pages': 1.0, ... |
+-------------------------------+
[183531 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# Define what is positive and negative sentiment

In [81]:
#ignore all 3*  reviews
products = products[products['rating']!= 3]

In [82]:
#positive sentiment = 4-star or 5-star reviews
products['sentiment'] = products['rating'] >= 4

In [44]:
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'recommend': 1.0, 'disapp... |     1     |
| {'quilt': 1.0, 'the': 1.0,... |     1     |
| {'tool': 1.0, 'clever': 1.... |     1     |
| {'rock': 1.0, 'many': 1.0,... |     1     |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |
| {'for': 1.0, 'barnes': 1.0... |     1     |
| {'right': 1.0, 'because': ... |     1     |
| {'like': 1.0, 'and': 1.0, ... |     1     |
| {'in': 1.0, 'pages': 1.0, ... |     1     |
| {'tracker': 1.0, 'now': 1.... |     1     |
+-------------------------------+-----------+
[166752 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# Train our sentiment classifier

In [83]:
train_data,test_data = products.random_split(.8,seed=0)

In [84]:
sentiment_model = turicreate.logistic_classifier.create(train_data,target='sentiment', features=['word_count'], validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 57356

Number of coefficients      : 57357

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 4        | 0.250000  | 0.918999     | 0.841421          | 0.840019            |

| 1         | 9        | 3.250000  | 2.075709     | 0.931359          | 0.911362            |

| 2         | 11       | 2.669978  | 2.673560     | 0.938650          | 0.916466            |

| 3         | 12       | 2.669978  | 3.074801     | 0.927417          | 0.901814            |

| 4         | 14       | 1.326072  | 3.665733     | 0.945200          | 0.918088            |

| 9         | 20       | 1.326072  | 5.693627     | 0.977654          | 0.917698            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

In [85]:
sentiment_model.evaluate(test_data)

{'accuracy': 0.9176975738650012,
 'auc': 0.9342357833151299,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      0       |        1        |  1397 |
 |      1       |        0        |  1344 |
 |      0       |        0        |  3931 |
 |      1       |        1        | 26632 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.951057941255245,
 'log_loss': 0.33047871872210066,
 'precision': 0.9501587641371436,
 'recall': 0.9519588218472976,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+--------------------+-------+------+
 | threshold |        fpr         |        tpr         |   p   |  n   |
 +-----------+--------------------+--------------------+-------+------+
 |    0.

### Accuracy of sentiment model 'accuracy': 0.9176975738650012

In [86]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

In [87]:
def word_count (word_dict, word):
    if not (word in word_dict):
        return 0
    return word_dict[word]

In [88]:
for word in selected_words :
    products[word] = products['word_count'].apply(lambda x: word_count(x,word))

# Question 1

In [89]:
for word in selected_words :
    print(word, products[word].sum())

awesome 3892.0
great 55791.0
fantastic 1664.0
amazing 2628.0
love 41994.0
horrible 1110
bad 4183
terrible 1146.0
awful 687
wow 425
hate 1107


In [90]:
selected_words_model = turicreate.logistic_classifier.create(train_data,target='sentiment', features= selected_words, validation_set=test_data)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133448

Number of classes           : 2

Number of feature columns   : 11

Number of unpacked features : 11

Number of coefficients      : 12

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+--------------+-------------------+---------------------+

| 1         | 2        | 0.223314     | 0.847401          | 0.845874            |

| 2         | 3        | 0.332410     | 0.847514          | 0.846085            |

| 3         | 4        | 0.441780     | 0.847626          | 0.846115            |

| 4         | 5        | 0.545102     | 0.847708          | 0.846385            |

| 5         | 6        | 0.658063     | 0.847708          | 0.846385            |

| 6         | 7        | 0.779738     | 0.847708          | 0.846385            |

+-----------+----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

# Question 2


In [127]:
sentiment_model.coefficients.sort('index').print_rows(num_rows=1000, num_columns=4)
##sentiment_model.coefficients.print_rows(num_rows=1000, num_columns=4)

+-------------+----------------------+-------+-----------------------+-----+
|     name    |        index         | class |         value         | ... |
+-------------+----------------------+-------+-----------------------+-----+
| (intercept) |         None         |   1   |   1.240508137457956   | ... |
|  word_count |          0           |   1   |  -0.7701126717681165  | ... |
|  word_count |          00          |   1   |  -0.9437426024788287  | ... |
|  word_count |         000          |   1   |  -0.7176877760475436  | ... |
|  word_count |         0001         |   1   |   2.4808091275882385  | ... |
|  word_count |        000ft         |   1   |   2.228440789477914   | ... |
|  word_count |         001          |   1   |   -7.906440589853358  | ... |
|  word_count |        001cm         |   1   |   0.3334804853930616  | ... |
|  word_count |         00am         |   1   |   0.6832922895372402  | ... |
|  word_count |      00amcreepy      |   1   |   -5.822364509703614  | ... |

# Question 3

In [118]:
selected_words_model.coefficients.sort('value' , ascending= False)

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 12

Data:
+-------------+-------+-------+----------------------+----------------------+
|     name    | index | class |        value         |        stderr        |
+-------------+-------+-------+----------------------+----------------------+
|     love    |  None |   1   |  1.3592688669224757  | 0.028068300152099997 |
| (intercept) |  None |   1   |  1.3365913848878148  | 0.008929969787657532 |
|   awesome   |  None |   1   |  1.1335346660341103  | 0.08399643983187567  |
|   amazing   |  None |   1   |  1.1000933113659914  | 0.09954776260465965  |
|  fantastic  |  None |   1   |  0.8858047568813963  | 0.11167591293399713  |
|    great    |  None |   1   |  0.8630655001195999  | 0.018955052444380473 |
|     wow     |  None |   1   | -0.00953823606771224 |  0.1604641122471162  |
|     bad     |  None |   1   | -0.9914778800650894  | 0.03848428664699065  |
|     hate    |  None |   1   | -1.3484407222463402  | 0.07715698604297318  |
|    awful    |  None |   1   |  -2.05290820403138   |  0.1009973543525922  |
+-------------+-------+-------+----------------------+----------------------+
[12 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [92]:
selected_words_model.evaluate(test_data)

{'accuracy': 0.8463848186404036,
 'auc': 0.6936022046674926,
 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 4
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      1       |        0        |  159  |
 |      0       |        0        |  371  |
 |      0       |        1        |  4957 |
 |      1       |        1        | 27817 |
 +--------------+-----------------+-------+
 [4 rows x 3 columns],
 'f1_score': 0.9157860082304526,
 'log_loss': 0.3962265467087835,
 'precision': 0.8487520595594068,
 'recall': 0.9943165570488991,
 'roc_curve': Columns:
 	threshold	float
 	fpr	float
 	tpr	float
 	p	int
 	n	int
 
 Rows: 100001
 
 Data:
 +-----------+--------------------+-----+-------+------+
 | threshold |        fpr         | tpr |   p   |  n   |
 +-----------+--------------------+-----+-------+------+
 |    0.0    |        1.0         | 1.0 | 27976 | 532

### Answer is 0.841 to 0.871

# Calculate accuracy of major classifier

In [94]:
total_positive = 159 + 27817

In [98]:
accuracy = total_positive/test_data.num_rows()

In [99]:
accuracy

0.8400192169108815

In [100]:
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']

In [101]:
diaper_champ_reviews

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	predicted_sentiment	float
	awesome	float
	great	float
	fantastic	float
	amazing	float
	love	float
	horrible	int
	bad	int
	terrible	float
	awful	int
	wow	int
	hate	int

Rows: Unknown

Data:
+-------------------------+-------------------------------+--------+
|           name          |             review            | rating |
+-------------------------+-------------------------------+--------+
| Baby Trend Diaper Champ | Ok - newsflash.  Diapers a... |  4.0   |
| Baby Trend Diaper Champ | My husband and I selected ... |  1.0   |
| Baby Trend Diaper Champ | Excellent diaper disposal ... |  5.0   |
| Baby Trend Diaper Champ | We love our diaper champ. ... |  5.0   |
| Baby Trend Diaper Champ | Two girlfriends and two fa... |  5.0   |
| Baby Trend Diaper Champ | I waited to review this un... |  4.0   |
| Baby Trend Diaper Champ | I have had a diaper genie ... |  1.0   |
| Baby Trend Diaper Champ | I originally put this item... |  5.0   |
| Baby Trend Diaper Champ | I am so glad I got the Dia... |  5.0   |
| Baby Trend Diaper Champ | We had 2 diaper Genie's bo... |  4.0   |
+-------------------------+-------------------------------+--------+
+-------------------------------+-----------+-----------------------+---------+
|           word_count          | sentiment |  predicted_sentiment  | awesome |
+-------------------------------+-----------+-----------------------+---------+
| {'convenient': 1.0, 'more'... |     1     |   0.9950122935568644  |   0.0   |
| {'system': 1.0, 'try': 1.0... |     0     | 5.901414536611702e-13 |   0.0   |
| {'nose': 1.0, 'for': 2.0, ... |     1     |   0.9999996193960344  |   0.0   |
| {'out': 1.0, 'pull': 1.0, ... |     1     |   0.9999567191544834  |   0.0   |
| {'winter': 1.0, 'outside':... |     1     |   0.9999997013199228  |   0.0   |
| {'mom': 1.0, 'huge': 1.0, ... |     1     |   0.9999920758179403  |   0.0   |
| {'yuck': 1.0, 'clean': 1.0... |     0     |   0.0017908398242359  |   0.0   |
| {'price': 1.0, 'suggestion... |     1     |   0.9999999999465673  |   0.0   |
| {'best': 1.0, 'that': 1.0,... |     1     |   0.9801174413564645  |   0.0   |
| {'no': 1.0, 'regular': 1.0... |     1     |   0.9999454483918991  |   0.0   |
+-------------------------------+-----------+-----------------------+---------+
+-------+-----------+---------+------+----------+-----+----------+-------+-----+
| great | fantastic | amazing | love | horrible | bad | terrible | awful | wow |
+-------+-----------+---------+------+----------+-----+----------+-------+-----+
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 1.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   1.0   | 0.0  |    1     |  0  |   0.0    |   0   |  1  |
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  1  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 0.0  |    0     |  0  |   0.0    |   0   |  0  |
|  0.0  |    0.0    |   0.0   | 2.0  |    0     |  0  |   0.0    |   0   |  0  |
+-------+-----------+---------+------+----------+-----+----------+-------+-----+
+------+
| hate |
+------+
|  0   |
|  0   |
|  0   |
|  0   |
|  0   |
|  0   |
|  0   |
|  0   |
|  0   |
|  0   |
+------+
[? rows x 17 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [102]:
diaper_champ_reviews['predicted_sentiment'] = sentiment_model.predict(diaper_champ_reviews, output_type = 'probability')

In [103]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [104]:
diaper_champ_reviews[0]

{'name': 'Baby Trend Diaper Champ',
 'review': "I read a review below that can explain exactly what we experienced. We've had it for 16 months and it has worked wonderful for us. No smells, change it out once a week, easy to clean. Then a diaper snagged this foam material in the head part, so I pulled the rest of the foam out. Big mistake!!! Now it can no loner retain the stinkiness and we're looking for a replacement. Be careful of overloading and never take out that foam piece that is cushioned between pieces. I have figured out that it is key to keeping the stink out.",
 'rating': 4.0,
 'word_count': {'key': 1.0,
  'have': 1.0,
  'pieces': 1.0,
  'between': 1.0,
  'cushioned': 1.0,
  'piece': 1.0,
  'take': 1.0,
  'overloading': 1.0,
  'be': 1.0,
  'looking': 1.0,
  're': 1.0,
  'stinkiness': 1.0,
  'retain': 1.0,
  'now': 1.0,
  'wonderful': 1.0,
  'worked': 1.0,
  '16': 1.0,
  'and': 3.0,
  'months': 1.0,
  've': 1.0,
  'in': 1.0,
  'us': 1.0,
  'i': 3.0,
  'experienced': 1.0,
  '

In [106]:
diaper_champ_reviews['predicted_sentiment2'] = selected_words_model.predict(diaper_champ_reviews, output_type = 'probability')

In [111]:
diaper_champ_reviews = diaper_champ_reviews.sort('predicted_sentiment', ascending=False)

In [112]:
diaper_champ_reviews[0]

{'name': 'Baby Trend Diaper Champ',
 'review': "I read a review below that can explain exactly what we experienced. We've had it for 16 months and it has worked wonderful for us. No smells, change it out once a week, easy to clean. Then a diaper snagged this foam material in the head part, so I pulled the rest of the foam out. Big mistake!!! Now it can no loner retain the stinkiness and we're looking for a replacement. Be careful of overloading and never take out that foam piece that is cushioned between pieces. I have figured out that it is key to keeping the stink out.",
 'rating': 4.0,
 'word_count': {'key': 1.0,
  'have': 1.0,
  'pieces': 1.0,
  'between': 1.0,
  'cushioned': 1.0,
  'piece': 1.0,
  'take': 1.0,
  'overloading': 1.0,
  'be': 1.0,
  'looking': 1.0,
  're': 1.0,
  'stinkiness': 1.0,
  'retain': 1.0,
  'now': 1.0,
  'wonderful': 1.0,
  'worked': 1.0,
  '16': 1.0,
  'and': 3.0,
  'months': 1.0,
  've': 1.0,
  'in': 1.0,
  'us': 1.0,
  'i': 3.0,
  'experienced': 1.0,
  '